src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -25,52 +26,49 @@ public async IAsyncEnumerable<string> GetFeatureNamesAsync()
         {
             if (_featureNames == null)
             {
-                var featureNames = new List<string>();
-
-                await foreach (string featureName in _featureManager.GetFeatureNamesAsync().ConfigureAwait(false))
-                {
-                    featureNames.Add(featureName);
-                }
-
-                _featureNames = featureNames;
+                _featureNames = await CreateListAsync(_featureManager).ConfigureAwait(false);
             }
 
             foreach (string featureName in _featureNames)
             {
                 yield return featureName;
             }
-        }
 
-        public async Task<bool> IsEnabledAsync(string feature)
-        {
-            //
-            // First, check local cache
-            if (_flagCache.ContainsKey(feature))
+            static async ValueTask<List<string>> CreateListAsync(IFeatureManager featureManager)
             {
-                return _flagCache[feature];
-            }
-
-            bool enabled = await _featureManager.IsEnabledAsync(feature).ConfigureAwait(false);
+                var featureNames = new List<string>();
 
-            _flagCache[feature] = enabled;
+                await foreach (string featureName in featureManager.GetFeatureNamesAsync().ConfigureAwait(false))
+                {
+                    featureNames.Add(featureName);
+                }
 
-            return enabled;
+                return featureNames;
+            }
         }
 
-        public async Task<bool> IsEnabledAsync<TContext>(string feature, TContext context)
+        public Task<bool> IsEnabledAsync(string feature)
         {
-            //
-            // First, check local cache
-            if (_flagCache.ContainsKey(feature))
-            {
-                return _flagCache[feature];
-            }
-
-            bool enabled = await _featureManager.IsEnabledAsync(feature, context).ConfigureAwait(false);
-
-            _flagCache[feature] = enabled;
+#if NETSTANDARD2_0
```

Comment:
Adding ifdefs increases the code complexity. I'm not sure it's worth it here. I'd opt for just sticking with the NETSTANDARD2_0 implementation.

Reply:
With NETSTANDARD2_0 we have repeated allocations for the delegate, that the overload with state (starting from NS 2.1) can avoid. See this [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6AYQHsAbauAYxgEtKA7KCt+gVwQTlYwA3LgLEMAFhF4iJAKzTcrCGDhQADhHpxCAWSb0ElKJQBmMUgDE4EGLzi6IygOZxVg3AG9chX8TSE1rb2ji6IAMrK6lAAFpQwAPoAcuFoRCCEAJJBdvyhEK4IkRDRcTA+ft44fjWE6ghMAG62OvwQACZs1ACeWTkhTgWIhAmmNrkOg4XStX71TS2EbZ2sPYRUrDx8AjAAIgbMbBAI3QA8JPgANMQAbKfAlDQAfE8jptQF5FoxOgC8hKw4AB3dZcXj8QT7RgsZQnc4YK63e6PagvAAUAEoZrNCBVZigAMxIh7PLJQACiymAdHaAEEoN1NmiLoQxsF+Bi8bUqjjaigAOxvD7OL70H6kADicBgAHkELT2u00WyJtdjs5CL9XqNxgMwghSJkKVSafTGfQ0eqMViuTUAL5ch3VPxc9CBXV5KYRKKxeLJcgEjCEDLZD2TfXFUrxLk8nHzZrwJY2FZrUPs8NDBBvMP5aa2uYNBOtZNdXobLYQvYHGHHM4Xa4oO4k1Ha96fb5/AHA0GbcE7KGHWF1hENpso9E252zfO+QnE8dkykQalwOkMpkslX2TlT7kzvmC0bC0XiqWy+WKtH73lb/iXa84y0IZzXUxgDGa15vw3G5em9cWla967ryNQ6umuaIJOvJOvaXKukSTCCIgphaDoaYTJBRQ+mUwZ9DmXoIDGjrwbuc5IfACCodo+EQYRxEgX4jbIqSRpLiua7msyCKsmG0F+E6TpAA=).

I believe it's worth it. 
Say it's used in a controller and checked for each request. So we can save #requests delegate allocations.

Reply:
Can't we avoid having the ifdefs intermingled with our business logic here by having an extension method with the 3.1 signature and having the ifdef there. I don't expect any unnecessary closures as a result. If we can do that I'm fine with it.

Reply:
Then again, optimizing for perf without numbers doesn't sound like the best strategy. For this it's hard to say how much we'd be saving in any case without numbers.

Reply:
> optimizing for perf without numbers 

The whole rationale of this API on ConcurrentDictionary is about saving the allocation of the delegate and potentially the closure. I don't need numbers for this, and won't do any benchmarks. So much discussion for this change...and the code reads quite nice with these two #ifdefs (IMHO). An extension method will just complicate things, as a) where to place it and b) it just creates more complicate closures.

I'll leave this PR as is. If it doesn't meet the bar, so please close it. Sorry that I have to say it that way.

Reply:
I stay away from ifdef without a strong reason to do so since I favor code maintenance and readability even though it may provide an optimization here for netcoreapp3.1. I would opt to remove it. It may just be me. @MSGaryWang @zhenlan for their insight.

Reply:
What do you think about https://github.com/gfoidl/FeatureManagement-Dotnet/commit/aaa47c3ab3f374ffedfd46706d5d02e939fad770
It's just one #ifdef, by passing `null` in as context (the FeatureManager does the same, so this has no side-effects).

Reply:
I see, here's my thoughts.
* If we are going to use ifdef I certainly do like the idea of limiting it to one place instead of multiple.
* This does have side-effects. The code no longer behaves the same way. [FeatureManager](https://github.com/microsoft/FeatureManagement-Dotnet/blob/2240941480f833fea27601368eb910260ce71966/src/Microsoft.FeatureManagement/FeatureManager.cs#L45) has different code paths when `IsEnabledAsync` is called with or without a context. Even if `null` is passed.
* I still would opt out of adding an ifdef in general, but would like to defer to one of the mentioned collaborators.